src/index.js

```diff
@@ -18,201 +15,227 @@ const logger = {
+                }
+                else {
+                    try {
+                        body = JSON.parse(body);
+                        if (!body || !body.authenticationSuccess || !body.user) {
+                            if (typeof(body.loginForm) === "string" && body.loginForm.includes('window.renderCaptcha()')) {
+                                errType = 'USER_ACTION_NEEDED.CAPTCHA';
+                            }
+                            else {
+                                errType = errors.LOGIN_FAILED;
+                            }
                         }
                     }
+                    catch (e) {
+                        console.log(e)
+                        logger.error("Cannot parse response");
+                        errType = errors.LOGIN_FAILED;
+                    }
                 }
-                catch (e) {
-                    logger.error("Cannot parse response");
-                    errType = errors.LOGIN_FAILED;
-                }
-            }
 
-            if (errType) {
-                logger.error("Signin failed");
-                return reject(new Error(errType));
-            }
+                if (errType) {
+                    logger.error("Signin failed");
+                    return reject(new Error(errType));
+                }
 
-            let cookie = `ID=${body.user.Id}&KEY=${body.user.AuthenticationCode}`;
-            j.setCookie(`Customer=${cookie}`, "https://secure.materiel.net");
+                let cookie = `ID=${body.user.Id}&KEY=${body.user.AuthenticationCode}`;
+                this._jar.setCookie(`Customer=${cookie}`, "https://secure.materiel.net");
 
-            logger.info("Logged in successfully");
-            return resolve();
+                logger.info("Logged in successfully");
+                return resolve();
+            });
         });
-    });
-}
+    }
 
-function fetchBillsPeriods() {
-    const billsOptions = {
-        method: "GET",
-        ecdhCurve: "auto",
-        jar: j,
-        url: `${baseURL}/Orders/CompletedOrdersPeriodSelection`
-    };
-
-    return new Promise((resolve, reject) => {
-        logger.info("Fetching bills periods");
-        request(billsOptions, (err, res, body) => {
-            if (!err) {
-                try {
-                    body = JSON.parse(body);
-                }
-                catch (e) {
-                    err = new Error("Could not parse bills periods list");
+    async fetchBillsPeriods() {
+        const billsOptions = {
+            method: "GET",
+            ecdhCurve: "auto",
+            url: `${baseURL}/Orders/CompletedOrdersPeriodSelection`
+        };
+
+        return new Promise((resolve, reject) => {
+            logger.info("Fetching bills periods");
+            this.request(billsOptions, (err, res, body) => {
+                if (!err) {
+                    try {
+                        if (typeof(body) === "string") {
```

Comment:
Ajout de ce test, ici l'objet json est obtenu directement plus besoin de le parser normalement.

Reply:
Could you add a comment please?